node/src/network/computation.rs

```diff
@@ -0,0 +1,85 @@
+use super::NetworkTaskChannel;
+use crate::tracking;
+use std::future::Future;
+
+/// Interface for a computation that is leader-centric:
+///  - If any follower's computation returns error, it automatically sends an Abort message to
+///    the leader, causing the leader to fail as well.
+///  - If the leader's computation returns error, it automatically sends an Abort message to
+///    all followers, causing their computation to fail as well.
+///
+/// If leader_waits_for_success returns true, then additionally:
+///  - Followers who succeed send a Success message to the leader.
+///  - The leader will wait for all Success messages before returning.
+///
+/// The leader_waits_for_success is for asset generation, where the owner of the asset wants
+/// to only mark it as completed when all followers have persisted their share of the asset.
+#[async_trait::async_trait]
+pub trait MpcLeaderCentricComputation<T>: 'static {
+    /// Performs the computation itself, without worrying about failure propagation or
+    /// waiting for success of followers.
+    async fn compute(self, channel: &mut NetworkTaskChannel) -> anyhow::Result<T>;
+    fn leader_waits_for_success(&self) -> bool;
+}
+
+pub trait MpcLeaderCentricComputationExt<T> {
+    fn perform_leader_centric_computation(
+        self,
+        channel: NetworkTaskChannel,
+        timeout: std::time::Duration,
+    ) -> impl Future<Output = anyhow::Result<T>> + 'static;
+}
```

Comment:
Why is this a separate trait instead of being part of `MpcLeaderCentricComputation`?

Reply:
I always feel that inserting a default implementation in a trait just to provide auxiliary utility is a bit weird because it doesn't quite belong to the trait and becomes override-able. The extension syntax is verbose on the other hand. I don't have a strong preference really.

Reply:
I see. I prefer having a default implementation in a trait. It is simpler and easier to read